content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,336 @@
+```
+
+The same should happen if the `User` had a field with a [deprecated name](https://www.scala-lang.org/api/current/scala/deprecatedName.html).
+
+## Order of name and term
+
+Normally, an equal sign assigns the result of the right side to the left side. With the proposed syntax, that's not the case. However, for most, if not all, people that's the correct order.
+
+This order seems to require a single look ahead in the parser.
+
+## Ignoring of unused parameters
+
+To allow case class to become more extensible, all unused parameters should be ignored. So when a field is added to the cases class old patterns stay source compatible.
+
+```scala
+  case User(city = "Paris") => // is the same as
+  case User(_, _, "Paris") =>
+```
+
+TODO: Find consent here
+
+## Syntax
+
+## Desugaring
+
+> One important principle of Scala’s pattern matching design is that case classes should be abstractable. I.e. we want to be able to represent the abilities of a case class without exposing the case class itself. That also allows code to evolve from a case class to a regular class or a different case class while maintaining the interface of the old case class. [Martin Odersky](https://contributors.scala-lang.org/t/pattern-matching-with-named-fields/1829/52)
+
+* Allow the same feature of case classes:
+  * Provide arbitrary names
+  * Enable [`@deprecatedName`](https://www.scala-lang.org/api/current/scala/deprecatedName.html), which leads to multiple names for a single field. Using both should be an error.
+* Maybe provide additional named fields. For example for our `User`:
+
+  ```scala
+  case User(ageInDays = 3650) => ...
+  ```
+
+  But this could be an awful feature.
+* Maybe leave fields without names
+
+# Encoding
+
+This SIP proposes basically to use [shapeless records](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#extensible-records) for the specific purpose of pattern matching.
+For an incomplete list of alternatives, see [alternative desugaring](#alternative-desugaring) below.
+
+Pro:
+
+* with shapeless records we have a good understanding what we are doing, however we need to tweak the encoding a bit
+* with some helper definitions, the usage could be quite short
+* reuse possible
+
+Con:
+
+* opens a can of worms, as this encoding could be useful for other requested features like [named tuples](named-tuple), which would require more thought-out design
+* no support for `@deprecatedName`
+* identifiers are represented with string literals
+* no meaningful way to use names within variadic patterns 
+  (the encoding enforces to use the names in order. A name like `last` in `Seq` could be implemented with this encoding.)
+
+Example for an extractor using option and tuple:
+
+```scala
+object User:
+  def unapply(user: User): Option[
+    (String, Int, String) & 
+    { type Names = ("name", "age", "city") }
+  ] = Some((user.name, user.age, user.city)).asInstanceOf[
```

Comment:
I think this implementation scheme is unacceptably complex. If want to make named-based pattern matching universally available, there should be no penalty to pay for writing extractors. In particular I would hope that the scheme we come with:

 - needs no `asInstanceOf`
 - needs no separate string-based field tuple

Reply:
This scheme isn't a starter, but my third attempt. (An attempt of a pun)

At least `asInstanceOf` can be circumvented with some helpers:

```scala
def unapply(user: User): (String, Int, String) & { type Names = ("name", "age", "city") } =
  ("name", user.name) +:: ("age", user.age) +:: ("city", user.city) +:: empty
```

(Full example here: [runNamedPatternMatching.scala](https://github.com/lampepfl/dotty/blob/01c79b1ca942550ac865cd3cdf267c2c7ec93b79/tests/run/runNamedPatternMatching.scala#L29))

Up to now I've seen two options to encode the names of the pattern in the return type of the extractor: case classes, as you describe below and the fields of the object returned by the extractor (is also described [above by sjrd](https://github.com/scala/improvement-proposals/pull/44#issuecomment-1191235753)). The first one was ruled out very early and the the second one would be bigger undertaking. I'm open other encodings.